### ISSUE TITLE
UriBuilder.fromResource and interfaces

### ISSUE THREAD

**Issue by mkarg** [eclipse-ee4j/jaxrs-api#790]
Apparently Jersey just checks the resource class itself, but not its interfaces, for `@Path`. The spec/javadocs are ambiguous here, as it is neither enforced for forbidden to check the interfaces of the root resource class.

From the view of a user, it would be great if all implementations MUST check all interfaces of the root resource class, not just the class itself.

Do we need to improve the javadocs/spec to have an unambiguous clarification whether this is enforced/forbidden?

@eclipse-ee4j/eclipse-jaxrs Opinions?

**Comment by chkal** [comment]
I think JAX-RS should use the same annotation inheritance rules for `UriBuilder.fromResource` as for the server-side (which are described in section 3.6 of the spec). This would be most consistent.

**Comment by mkarg** [comment]
@eclipse-ee4j/eclipse-jaxrs Vendors, what do you think?

**Comment by mkarg** [comment]
@jansupol @andymc12 @asoldano Can you please post your opinion on this question?

**Comment by andymc12** [comment]
I don't have any strong opinions here.  FWIW, the spec seems clear to me that only annotations on the resource _class_ are portable, but there is text to indicate that other implementors are free to change behavior.

**Comment by mkarg** [comment]
It's time to find an agreement. Proposal:

- There is no need to change neither Spec nor Javadocs.
- `UriBuilder.fromResource` MUST respect interfaces w.r.t the rules outlined by chapter 3.6 of the spec.

Please VOTE now. :-)

**Comment by ronsigal** [comment]
I agree with @chkal: "JAX-RS should use the same annotation inheritance rules for UriBuilder.fromResource as for the server-side (which are described in section 3.6 of the spec)".  Given that Section 3.6 says, "inheritance of class or interface annotations is not supported", I think that "Create a new instance representing a relative URI initialized from a root resource class" unambiguously says that only the root resource class should be examined.

So, 

 * I agree with "There is no need to change neither Spec nor Javadocs., but
 * I don't understand "UriBuilder.fromResource MUST respect interfaces w.r.t the rules outlined by chapter 3.6 of the spec".

**Comment by mkarg** [comment]
My proposal is a spec change. When designing applications it is really not nice that annotations have to get repeated in the implementing class. Why don't we change the spec and do allow inheritance of annotations from interfaces? Are there technical reasons against this? It would make complex code cleaner to read due to DRY.

**Comment by andymc12** [comment]
Sorry, I think I'm confused on what is being proposed.  As I understand it, Jersey's `UriBuilder.fromResource(SomeResourceClass.class)` implementation does not return URI paths from `@Path` annotations on super classes or implemented interfaces.  But section 3.6 says that class-level annotations are not supported:
> JAX-RS annotations may be used on the methods and method parameters of a super-class or an implemented interface. Such annotations are inherited by a corresponding sub-class or implementation class method provided that the method and its parameters do not have any JAX- RS annotations of their own. Annotations on a super-class take precedence over those on an implemented interface. The precedence over conflicting annotations defined in multiple implemented interfaces is implementation specific. Note that inheritance of class or interface annotations is not supported.

So it sounds like Jersey's implementation already respects the rules in 3.6.  

Is the proposal to add some clarifying Javadoc text to `UriBuilder.fromResource(...)` to indicate that class-level path annotation inheritance is not supported?  Or is the proposal to change section 3.6 so that class-level annotations is now supported (both for server side processing and for the `UriBuilder.fromResource(...)` method?  Or something else?

**Comment by mkarg** [comment]
> Is the proposal to add some clarifying Javadoc text to `UriBuilder.fromResource(...)` to indicate that class-level path annotation inheritance is not supported? Or is the proposal to change section 3.6 so that class-level annotations is now supported (both for server side processing and for the `UriBuilder.fromResource(...)` method? Or something else?

My proposal actually is to convince everybody that UriBuilder.fromResource() now supports class-level annotation inheritance. This is what feels most intuitive to the average programmer in comparison to the current situation of repeating annotations.

**Comment by chkal** [comment]
I agree that the annotation inheritance rules described in section 3.6 are a bit "special" and may not be very intuitive. But changing them would be challenging in regards to backwards compatibility. Unfortunately.

**Comment by ronsigal** [comment]
I've never understood why the current restrictions on annotation inheritance exist. Was there a compelling technical reason? And then we have to say explicitly that CDI annotation inheritance is different. Would it be possible to just adopt the CDI inheritance rules?

**Comment by chkal** [comment]
> I've never understood why the current restrictions on annotation inheritance exist. Was there a compelling technical reason?

I also wondered what the reason was to define it this way. 

@spericas Any idea?

**Comment by spericas** [comment]
I don't know the history of that, unfortunately. They are a bit odd for sure. We should definitely review this in the next major release.
